core/drivers/stm32_rtc.c

```diff
@@ -630,9 +803,227 @@ TEE_Result stm32_rtc_driver_is_initialized(void)
 	return TEE_ERROR_DEFER_DRIVER_INIT;
 }
 
+static TEE_Result stm32_rtc_read_alarm(struct rtc *rtc,
+				       struct optee_rtc_alarm *alarm)
+{
+	struct optee_rtc_time *alarm_tm = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_time current_tm = { };
+	vaddr_t rtc_base = get_base();
+	uint32_t alrmar = io_read32(rtc_base + RTC_ALRMAR);
+	uint32_t cr = io_read32(rtc_base + RTC_CR);
+	uint32_t status = io_read32(rtc_base + RTC_SR);
+
+	alarm_tm = &alarm->time;
+
+	res = stm32_rtc_get_time(rtc, &current_tm);
+	if (res)
+		return res;
+
+	alarm_tm->tm_year = current_tm.tm_year;
+	alarm_tm->tm_mon = current_tm.tm_mon;
+	alarm_tm->tm_mday = ((alrmar & RTC_ALRMXR_DATE_UNITS_MASK) >>
+			    RTC_ALRMXR_DATE_UNITS_SHIFT) +
+			    ((alrmar & RTC_ALRMXR_DATE_TENS_MASK) >>
+			    RTC_ALRMXR_DATE_TENS_SHIFT) * 10;
+	alarm_tm->tm_hour = ((alrmar & RTC_ALRMXR_HOUR_UNITS_MASK) >>
+			    RTC_ALRMXR_HOUR_UNITS_SHIFT) +
+			    ((alrmar & RTC_ALRMXR_HOUR_TENS_MASK) >>
+			    RTC_ALRMXR_HOUR_TENS_SHIFT) * 10;
+	alarm_tm->tm_min = ((alrmar & RTC_ALRMXR_MIN_UNITS_MASK) >>
+			    RTC_ALRMXR_MIN_UNITS_SHIFT) +
+			   ((alrmar & RTC_ALRMXR_MIN_TENS_MASK) >>
+			    RTC_ALRMXR_MIN_TENS_SHIFT) * 10;
+	alarm_tm->tm_sec = ((alrmar & RTC_ALRMXR_MIN_UNITS_MASK) >>
+			    RTC_ALRMXR_MIN_UNITS_SHIFT) +
+			   ((alrmar & RTC_ALRMXR_MIN_TENS_MASK) >>
+			    RTC_ALRMXR_MIN_TENS_SHIFT) * 10;
+
+	if (rtc_timecmp(alarm_tm, &current_tm) < 0) {
+		if (current_tm.tm_mon == 11) {
+			alarm_tm->tm_mon = 0;
+			alarm_tm->tm_year += 1;
+		} else {
+			alarm_tm->tm_mon += 1;
+		}
+	}
+
+	alarm->enabled = cr & RTC_CR_ALRAE;
+	alarm->pending = status & RTC_SR_ALRAF;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rtc_enable_alarm(struct rtc *rtc __unused, bool enabled)
+{
+	vaddr_t rtc_base = get_base();
+
+	stm32_rtc_write_unprotect();
+
+	if (enabled)
+		io_setbits32(rtc_base + RTC_CR, RTC_CR_ALRAIE | RTC_CR_ALRAE);
+	else
+		io_clrbits32(rtc_base + RTC_CR, RTC_CR_ALRAIE | RTC_CR_ALRAE);
+
+	stm32_rtc_clear_events(RTC_SCR_CALRAF);
+
+	stm32_rtc_write_protect();
+
+	return TEE_SUCCESS;
+}
+
+static void stm32_rtc_add_one_month(struct optee_rtc_time *tm)
+{
+	tm->tm_mon++;
+	if (tm->tm_mon > 11) {
+		tm->tm_mon = 0;
+		tm->tm_year++;
+	}
+
+	/* Saturate to the next month last day */
+	tm->tm_mday = MIN(tm->tm_mday,
+			  rtc_get_month_days(tm->tm_mon, tm->tm_year));
+}
+
+static TEE_Result stm32_rtc_valid_alarm_time(struct rtc *rtc,
+					     struct optee_rtc_time *tm)
+{
+	struct optee_rtc_time current_tm = { };
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_time *max = NULL;
+
+	/*
+	 * Assuming current date is M-D-Y H:M:S.
+	 * RTC alarm can't be set on a specific month and year.
+	 * So the valid alarm range is:
+	 *	M-D-Y H:M:S < alarm <= (M+1)-D-Y H:M:S
+	 */
+
+	res = stm32_rtc_get_time(rtc, &current_tm);
+	if (res)
+		return res;
+
+	/* Don't allow alarm to be set in the past. */
+	if (rtc_timecmp(&current_tm, tm) >= 0)
```

Comment:
I don't understand why we would allow to set an alarm at the exact same time? and not keep `>`

Reply:
If at exact time, the alarm would trigger 1 month later if I'm right. Is it what is expected?